tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/IntegerQuantificationOptimizer.scala

```diff
@@ -0,0 +1,88 @@
+ *
+ * @author
+ *   Jure Kukovec
+ */
+
+@Singleton
+class IntegerQuantificationOptimizer(tracker: TransformationTracker) extends TlaExTransformation {
+
+  private val intTag = Typed(IntT1())
+  private val boolTag = Typed(BoolT1())
+
+  override def apply(expr: TlaEx): TlaEx = {
+    transform(expr)
+  }
+
+  // This optimization only applies to Int, Nat and a..b sets (of integers)
+  // TODO: Unused ATM, but will be needed when plugging this transformation into preprocessing
+  private def isAllowedSet(s: TlaEx): Boolean = s match {
+    case ValEx(TlaIntSet | TlaNatSet)      => true
+    case OperEx(TlaArithOper.dotdot, _, _) => true
+    case _                                 => false
+  }
+
+  // Given a quantifier, returns the unbounded variant of the same quantifier.
+  def mkUnbounded(op: TlaOper): TlaOper = op match {
+    case TlaBoolOper.forall => TlaBoolOper.forallUnbounded
+    case TlaBoolOper.exists => TlaBoolOper.existsUnbounded
+    case _                  => op
+  }
+
+  // name >= 0
+  private def ge0(name: String): TlaEx =
+    OperEx(TlaArithOper.ge, NameEx(name)(intTag), ValEx(TlaInt(0))(intTag))(Typed(BoolT1()))
+
+  // a <= name <= b
+  private def inRange(name: String, lowerBound: TlaEx, upperBound: TlaEx): TlaEx = {
+    val lConstraint = OperEx(TlaArithOper.le, lowerBound, NameEx(name)(intTag))(boolTag)
+    val uConstraint = OperEx(TlaArithOper.le, NameEx(name)(intTag), upperBound)(boolTag)
+    OperEx(TlaBoolOper.and, lConstraint, uConstraint)(boolTag)
+  }
+
+  // (\A | \E) x \in S: P transforms into (\A | \E) x: Q
+  // For \E x \in S: P,
+  // Q := R /\ P, for some R
+  // and for \A x \in S: P,
+  // Q := R => P, for some R
+  private def andOrImpl(op: TlaOper): TlaBoolOper =
+    if (op == TlaBoolOper.exists) TlaBoolOper.and
+    else TlaBoolOper.implies
+
+  // Creates an expression (\A | \E) x: op(P2,P),
+  // where op is either `/\` or `=>`
+  private def modifyPred(
+      oper: TlaOper,
+      nameEx: TlaEx,
+      setMemberPred: TlaEx,
+      originalPred: TlaEx): TlaEx = {
+    val newPred = OperEx(andOrImpl(oper), setMemberPred, originalPred)(boolTag)
+    OperEx(mkUnbounded(oper), nameEx, newPred)(boolTag)
+  }
+
+  def transform: TlaExTransformation = tracker.trackEx {
+    // Assume typechecking succeeded -> name has type Int
+    case OperEx(oper @ (TlaBoolOper.forall | TlaBoolOper.exists), nameEx, ValEx(TlaIntSet), pred) =>
+      OperEx(mkUnbounded(oper), nameEx, pred)(boolTag)
+    case OperEx(oper @ (TlaBoolOper.forall | TlaBoolOper.exists), nameEx @ NameEx(name), ValEx(TlaNatSet), pred) =>
+      modifyPred(oper, nameEx, ge0(name), pred)
+    case OperEx(oper @ (TlaBoolOper.forall | TlaBoolOper.exists), nameEx @ NameEx(name),
+            OperEx(TlaArithOper.dotdot, lowerBound, upperBound), pred) =>
+      modifyPred(oper, nameEx, inRange(name, lowerBound, upperBound), pred)
+    case ex => ex
+  }
```

Comment:
NOTE: This comment is just a suggestion, and more to discuss the general way we break down this kind of logic than to insist on any particular change. If you don't think the suggestion helps clarify the logic, feel free to disregard.

This may come down to a matter of taste, but the way the logic of this function is currently factored is a bit confusing to me. I think the following things throw me off while reading:

- We end up pattern matching on the quantifiers in 3 different places
- The helper function `mkUnbounded` gets applied in two logically different places (once on line 74, and again on line 80) even tho the underlying logic of the transformations in each case is really quite similar.
- Instead of applying a few different transformations to construct the new data we need, and then constructing the transformation using that new data, we end up sort of (imo) "hiding" the sequence of transformations in nested function calls.

An alternative that aims to consolidate the related changes, to give what is (imo) a clearer and more concise diagram of the overall transformation being made, might look something like this:

```suggestion
  def transform: TlaExTransformation = tracker.trackEx {
    // Assume typechecking succeeded -> name has type Int
    case OperEx(oper @ (TlaBoolOper.forall | TlaBoolOper.exists), name, bound, pred) if isAllowedSet(bound) => {
      val (unboundedOper, outerPred) = oper match {
        case TlaBoolOper.forall => (TlaBoolOper.forallUnbounded, TlaBoolOper.implies)
        case TlaBoolOper.exists => (TlaBoolOper.existsUnbounded, TlaBoolOper.and)
      }
      val newPred = bound match {
        case ValEx(TlaIntSet) => pred
        case ValEx(TlaNatSet) => OperEx(outerPred, ge0(name), pred)
        case OperEx(TlaArithOper.dotdot, lowerbound, upperbound) =>
          OperEx(outerPred, inRange(name, lowerbound, upperbound), pred)
      }
      OperEx(unboundedOper, name, newPred)
    }
  }
```

And the `mkUnbounded` would also not be needed.

Reply:
Yeah, this is celeaner.